Marko support rerendering

DylanPiercey opened the pull request:
Issue:

Currently the Marko API for stories is to return a render result. This limits the ability for stories to be efficiently rerendered while maintaining state and also locks in users to a specific render API for Marko.

## What I did

This PR adds a deprecation warning for the old api of returning the `renderSync` result of a component, and now accepts an object result with the template to render and the input for the template similar to the `vue`, `angular`, `svelte` and some other implementations.

With this change I was able to keep track of the currently rendered component instance and only destroy the existing component if the constructor has been modified. This means that when using an addon like `knobs` the component is efficiently rerendered instead of destroyed. This also means that the entire lifecycle of the component is no longer rerun during updates which I think is more inline with what developers expect.

## How to test

I am unsure what the best way to test this is, however I have updated all examples I could find.

DylanPiercey commented:
I botched this PR, give me a minute to fix 😄

DylanPiercey committed:
feat: update Marko api to support rerendering

DylanPiercey commented:
Alright, I believe I've fixed everything. Happy to address any feedback.

DylanPiercey committed:
feat: support setting Marko template as a shared parameter

ndelangen commented:
@shilman seems really useful, can we push this into 5.2 still?

shilman commented on examples/marko-cli/src/stories/addon-knobs.stories.js:
```diff
@@ -5,16 +5,14 @@ export default {
   decorators: [withKnobs],
   parameters: {
-    component: Hello,
+    template: Hello,
```
component param is for addon-docs

DylanPiercey replied:
Thanks for the pointer! Could you share with me the docs for that config option? I’m new to using story book and want to make sure that Marko is accommodating as much of the ecosystem as possible 🙂.

shilman replied:
It hasn't made it into the main docs yet (coming soon!), but here's the best resource for now. https://docs.google.com/document/d/1un6YX7xDKEKl5-MVb-egnOYN8dynb5Hf7mq0hipk8JE/edit

DylanPiercey replied:
Excellent, thanks. Perhaps it would make sense to use `component` instead of `template` as a property for the story fns as well just to be consistent. I’ll update the PR shortly.

shilman commented on examples/marko-cli/src/stories/addon-actions.stories.js:
```diff
@@ -4,8 +4,11 @@ import Button from '../components/action-button/index.marko';
   title: 'Addons|Actions/Button',
   parameters: {
-    component: Button,
+    options: { panelPosition: 'right' },
```
component param is for addon-docs

shilman reviewed (changes requested):
A few minor things

DylanPiercey committed:
fix: add back component parameter option, rename template to component

DylanPiercey commented:
@shilman I've added back the `component` parameter option to the config. I've also changed the return value from the story functions to also expect a property called `component` instead of `template` since I think this makes the whole thing more cohesive.

I've it up such that if there is a `component` option in the parameter options that that will be used if no `component` option was set as the return from a story function. This I think makes it easier to define multiple stories for a single component with varying inputs. Hopefully overloading this parameter property as a shorthand is some cases is acceptable 😄

shilman commented:
@DylanPiercey Is this a breaking change?

DylanPiercey commented:
@shilman this should not be a breaking change. I added a deprecation message for the old behavior.

shilman committed:
Update addon-actions.stories.js
